# BÀI TẬP PHÂN LOẠI

## 📘 (1) Giới Thiệu
### 🦅 **BÀI TOÁN PHÂN LOẠI ĐẠI BÀNG MACBRIDE HAWKS BẰNG CÁC MÔ HÌNH PHÂN LOẠI**
****
### ℹ️ **Thông tin học viên**

Họ tên: Bùi Tấn Thuận  
MSHV: MIT24202  
Lớp: Học Máy - Machine Learning

### ℹ️ **Giới thiệu bài toán** <br>
Bài toán phân loại đại bàng **Fruit360** là bài toán phân loại với **3** loài chim sử dụng các mô hình học máy để phân loại. Trong quá trình thực hiện, bài toán cũng trình bày quá trình khám phá dữ liệu cũng như đánh giá và so sánh các loại mô hình.

### 🎯 **Mục tiêu** <br>
- Làm quen với quy trình đầy đủ của một bài toán phân loại trong học máy:
    - Cài đặt và Import thư viện.
    - Tải và tiền xử lý dữ liệu.
    - Khám phá dữ liệu (EDA).
    - Xây dựng và huấn luyện các mô hình phân loại.
    - Đánh giá và so sánh các mô hình
- Thực hành với tập dữ liệu khác với tập dữ liệu trong thư viện sklearn :
    - Với 3 loài chim săn mồi Red-tailed, Sharp-shinned, và Cooper’s hawks.
    - Nhãn gồm: tên loại đại bàng (Class)

### ℹ️ **Tập dữ liệu** <br>

- Nguồn gốc: Được thu thập bởi sinh viên và giảng viên tại Cornell College, tại điểm quan sát chim ở hồ MacBride, Iowa, Mỹ.
- Tập dữ liệu được tổ chức thành 1 file hawks_news.csv với:
    - Số mẫu: 447 cá thể chim
    - Số đặc trưng: 9 đặc trưng về sinh học của đại bàng
    - Trong đó đặc trưng 'Species' là tên loài

Trang chủ: https://www.kaggle.com/datasets/mathsian/three-hawk-species
****

## 🧩 (2) Cài đặt & Import thư viện

- Sử dụng các thư viện để hỗ trợ cho việc tiền xử lý, trực quan hóa, hiển thị, xử lý và đánh giá mô hình và dữ liệu.
****

In [1]:
import pandas as pd
import numpy as np
import plotly.express as px
import os
import zipfile
import subprocess
from sklearn.metrics import confusion_matrix

## 📥 (3) Tải dữ liệu
- Sử dụng Kaggle CLI API để tải dữ liệu về từ Kaggle
****

In [ ]:
# === BƯỚC 1: Cấu hình kaggle.json ===
kaggle_config_dir = os.path.expanduser("~/Desktop/MasterHK2/.kaggle")
os.environ['KAGGLE_CONFIG_DIR'] = kaggle_config_dir

# === BƯỚC 2: Thư mục lưu tập tin tải về ===
download_dir = os.path.expanduser("~/Desktop/MasterHK2/Dataset/hawks")
os.makedirs(download_dir, exist_ok=True)

# === BƯỚC 3: Tải dữ liệu từ Kaggle ===
zip_path = os.path.join(download_dir, "three-hawk-species.zip")
if not os.path.exists(zip_path):
    subprocess.run([
        "kaggle", "datasets", "download",
        "-d", "mathsian/three-hawk-species",
        "-p", download_dir
    ])
    print("📦 Đã tải về file three-hawk-species.zip")
else:
    print("✅ File three-hawk-species.zip đã tồn tại, bỏ qua bước tải.")

# === BƯỚC 4: Giải nén ===
extract_dir = os.path.join(download_dir, "three-hawk-species")
if not os.path.exists(extract_dir):
    with zipfile.ZipFile(zip_path, 'r') as zip_ref:
        zip_ref.extractall(extract_dir)
    print(f"✅ Đã giải nén dữ liệu vào: {extract_dir}")
else:
    print(f"📂 Dữ liệu đã được giải nén sẵn tại: {extract_dir}")

## 📊 (4) EDA - Khám phá dữ liệu
Trong phần này, chúng ta sẽ thực hiện một số phân tích sơ bộ về tập dữ liệu:

- Hiển thị sơ bộ 5 dòng ngẫu nhiên của tập dữ liệu.
- Hiển thị thông tin tổng quan.
- Hiển thị phân bố các mẫu từng lớp.

Các bước này giúp ta hiểu rõ hơn về dữ liệu trước khi đưa vào huấn luyện mô hình.
****

### 4.1. ▶️ Hiển thị sơ bộ 5 dòng ngẫu nhiên của tập dữ liệu
****

In [3]:
import pandas as pd
df = pd.read_csv('C:/Users/cva16/Desktop/MasterHK2/Dataset/hawks/hawks/hawks_new.csv')
df.sample(5)

,Year,Species,Wing,Weight,Tail,Hallux,Coopers?,Red-tailed?,Sharp-shinned?
63,1994,RT,380.0,985.0,223,29.0,0,1,0
369,2002,SS,159.0,90.0,130,11.5,0,0,1
326,2001,RT,414.0,1215.0,242,29.2,0,1,0
15,1992,SS,205.0,180.0,164,15.0,0,0,1
432,2003,SS,161.0,95.0,129,11.7,0,0,1


### 4.2. ▶️ Hiển thị thông tin tổng quan
****

In [4]:
from IPython.display import display, HTML

# Danh sách tên các lớp
class_names = sorted(df['Species'].unique())

# Tổng số mẫu
total_samples = len(df)

#Tổng giá trị null
total_nulls = df.isnull().sum().sum()

# Số cột hiển thị
columns = 3
colors = ['#e0f7fa', '#fbe9e7', '#f3e5f5']  # Màu nền xen kẽ

# Tạo HTML
html = f"""
<div style="padding: 14px; background-color: #f9f8e6; text-align: center; border-left: 5px solid #2ecc71;
            border-radius: 8px; font-family: sans-serif; font-size: 16px; color: #2c3e50; font-weight: bold;
            margin-bottom: 15px;">
    📊 Tổng số mẫu: {total_samples:,} | 🧩 Tổng số lớp: {len(class_names)} | Tổng giá trị Null: {total_nulls}
</div>

<h3 style='text-align: center; font-weight: bold; font-family: sans-serif; color: #2c3e50; margin-bottom: 20px;'>
  📚 Danh sách lớp
</h3>

<div style='display: grid; grid-template-columns: repeat({columns}, 1fr); gap: 10px;
            font-family: sans-serif; font-size: 14px;'>
"""

for i, name in enumerate(class_names):
    color = colors[i % columns]
    html += f"""
    <div style='padding: 8px 12px; background: {color}; border-radius: 6px; text-align: center;
                box-shadow: 1px 1px 3px rgba(0,0,0,0.05);'>
        {name}
    </div>
    """

html += "</div>"

display(HTML(html))

### 4.3. ▶️ Hiển thị phân bố các mẫu từng lớp
****

In [5]:
# Phân phối theo lớp
label_counts = df['Species'].value_counts()

# Vẽ biểu đồ phân phối theo lớp
fig = px.bar(
    label_counts,
    x=label_counts.index,
    y=label_counts.values,
    color=label_counts.index,
    text=label_counts.values,
    labels={
        'Label': 'Nhãn'
    },
    title='<b>Phân phối nhãn</b>'
    )
fig.update_layout(
    width=800,
    height=500,
    xaxis_title='Nhãn',
    yaxis_title='Số lượng',
    title={
    'y':0.95, # Adjust y position slightly for centering effect
    'x':0.5,  # Center the title horizontally
    'xanchor': 'center',
    'yanchor': 'top'
    },
    legend_title_text='Nhãn'
)
fig.show()

- Phân phối nhãn không đều:
    - Nhãn RT có số lượng khá lớn so với 2 lớp còn lại. Mô hình sẽ tập trung vào lớp này.
    - Nhãn CH có số lượng ít nhất. Khả năng sẽ bị lơ là trong lúc học.

## ⏳ (5) Tiền xử lý dữ liệu
****

### 5.1. ✂️ Xử lý các giá trị null/naN
- Loại bỏ các mẫu có các đặc trưng về nội dung và loại tin bị null.
****

In [6]:
df = df.dropna()

### 5.2. 🏷️ Gán nhãn & Mã hóa
Sử dụng LabelEncoder để gán nhãn và mã hóa.
****

In [7]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
df['label'] = le.fit_transform(df['Species'])

### 5.3. ✂️ Chia Train/Val/Test
Chia các tập Train/Val/Test theo tỉ lệ 70-15-15.
****

In [8]:
from sklearn.model_selection import train_test_split

# Chia dữ liệu thành Train/Val/Test

X = df.drop(columns=['Species'])
y = df['Species']

X_train, X_temp, y_train, y_temp = train_test_split(X, y, test_size=0.3, random_state=42, stratify=y)

X_val, X_test, y_val, y_test = train_test_split(X_temp, y_temp, test_size=0.5, random_state=42, stratify=y_temp)

In [9]:
# Hiển thị kích thước của các tập dữ liệu
sizes = [len(y_train), len(y_val), len(y_test)]
names = ['Train', 'Validation', 'Test']
fig_bar = px.bar(
    x=names, 
    y=sizes, 
    labels={'x': 'Dataset', 'y': 'Size'}, 
    title='<b>Kích thước các tập dữ liệu</b>',
    color=names,
    text_auto=True)
fig_bar.update_layout(
    width=800, 
    height=500,
    xaxis_title='Tập dữ liệu',
    yaxis_title='Kích thước',
    yaxis=dict(
        tickformat=',d'  # Hiển thị số nguyên, ví dụ: 14614 thay vì 14.6k
    ),
    legend_title_text='Tập dữ liệu',
    title={
    'y':0.95, # Adjust y position slightly for centering effect
    'x':0.5,  # Center the title horizontally
    'xanchor': 'center',
    'yanchor': 'top'
    },)
fig_bar.show()


# Calculate percentages
total = sum(sizes)
percentages = [f"{(s / total) * 100:.1f}%" for s in sizes]

# Create pie chart
fig_pie = px.pie(
    names=names,
    values=sizes,
    title='<b>Tỷ lệ phần trăm các tập dữ liệu</b>',
    hole=0.3  # Optional: makes it a donut chart
)
fig_pie.update_traces(
    textinfo='label+percent',
    hovertemplate='%{label}: %{value} mẫu (%{percent})'
)
fig_pie.update_layout(
    width=800, 
    height=500,
    legend_title_text='Tập dữ liệu',
    title={
        'y': 0.95,
        'x': 0.5,
        'xanchor': 'center',
        'yanchor': 'top'
    }
)
fig_pie.show()

- Tập dữ liệu được chia theo tỉ lệ ~ 70-15-15
    - Tập train chiếm ~70%. Phù hợp cho mô hình học.
    - Tập test và validation mỗi tập chiếm 15%. Phù hợp để đánh giá.

## 🏗️ (6) Xây dựng mô hình
- Định nghĩa các Pipeline.
- Định nghĩa các mô hình học máy.
****

In [10]:
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import GridSearchCV

from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.base import BaseEstimator


# Pipeline gồm bước chuẩn hóa và placeholder mô hình
pipe = Pipeline([
    ('scaler', StandardScaler()),
    ('model', BaseEstimator())  # placeholder, sẽ được thay bằng GridSearchCV
])

param_grid = [
    {
        'model': [GaussianNB()]
    },
    {
        'model': [SVC()],
        'model__C': [0.1, 1, 10],
        'model__kernel': ['linear', 'rbf']
    },
    {
        'model': [DecisionTreeClassifier()],
        'model__max_depth': [3, 5, None],
        'model__criterion': ['gini', 'entropy']
    },
    {
        'model': [RandomForestClassifier()],
        'model__n_estimators': [50, 100],
        'model__max_depth': [None, 5]
    },
    {
        'model': [MLPClassifier(max_iter=1000)],
        'model__hidden_layer_sizes': [(50,), (100,)],
        'model__activation': ['relu', 'tanh']
    }
]

## 🏋️ (7) Huấn luyện mô hình
- Sử dụng CrossValidation để duyệt qua từng mô hình với các tham số khác nhau.
****

In [11]:
from sklearn.model_selection import GridSearchCV

grid = GridSearchCV(pipe,
                    param_grid=param_grid,
                    cv=5,
                    scoring='accuracy',
                    refit=True,
                    n_jobs=-1,
                    return_train_score=True)

grid.fit(X_train, y_train)

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('scaler', StandardScaler()),
                                       ('model', BaseEstimator())]),
             n_jobs=-1,
             param_grid=[{'model': [GaussianNB()]},
                         {'model': [SVC()], 'model__C': [0.1, 1, 10],
                          'model__kernel': ['linear', 'rbf']},
                         {'model': [DecisionTreeClassifier()],
                          'model__criterion': ['gini', 'entropy'],
                          'model__max_depth': [3, 5, None]},
                         {'model': [RandomForestClassifier()],
                          'model__max_depth': [None, 5],
                          'model__n_estimators': [50, 100]},
                         {'model': [MLPClassifier(max_iter=1000)],
                          'model__activation': ['relu', 'tanh'],
                          'model__hidden_layer_sizes': [(50,), (100,)]}],
             return_train_score=True, scoring='accuracy')

## 🎯 (8) Đánh giá mô hình
****

In [16]:
# Trích mô hình tốt nhất từ pipeline
best_model = grid.best_estimator_.named_steps['model']

# In tên mô hình và thông tin chi tiết
print("✅ Best Pipeline:", grid.best_estimator_)
print("📦 Best Model:", best_model.__class__.__name__)
print("🎯 CV Accuracy:", grid.best_score_)
print("🔧 Best parameters:", grid.best_params_)

✅ Best Pipeline: Pipeline(steps=[('scaler', StandardScaler()), ('model', GaussianNB())])
📦 Best Model: GaussianNB
🎯 CV Accuracy: 1.0
🔧 Best parameters: {'model': GaussianNB()}


- Mô hình tốt nhất là mô hình GaussianNB
- Với pipeline : (StandardScaler, GaussianNB)

## ✅ (9) Dự đoán và đánh giá trên tập Test

In [13]:
from sklearn.metrics import classification_report, accuracy_score

# Dự đoán và đánh giá trên test set
y_test_pred = grid.predict(X_test)

print("\n📊 Test set classification report:")
print(classification_report(y_test, y_test_pred))

print("🎯 Test Accuracy:", accuracy_score(y_test, y_test_pred))


📊 Test set classification report:
              precision    recall  f1-score   support

          CH       1.00      1.00      1.00         5
          RT       1.00      1.00      1.00        43
          SS       1.00      1.00      1.00        20

    accuracy                           1.00        68
   macro avg       1.00      1.00      1.00        68
weighted avg       1.00      1.00      1.00        68

🎯 Test Accuracy: 1.0


- Mô hình GaussianNB có accuracy là 100%

## 🏁 (10) Ma trận nhầm lẫn

In [17]:
def plot_confusion_matrix_plotly(y_true, y_pred, model_name):
    cm = confusion_matrix(y_true, y_pred)
    cm_df = pd.DataFrame(cm, index=le.classes_, columns=le.classes_)

    
    fig = px.imshow(cm_df,
                    text_auto=True,
                    color_continuous_scale='Blues',
                    labels=dict(x="Predicted", y="Actual", color="Count"),
                    title=f"📊 Confusion Matrix – {model_name}")
    fig.update_layout(
        template='plotly_white',
        width = 800,
        height = 500,)
    fig.show()
    
plot_confusion_matrix_plotly(y_test, y_test_pred, best_model.__class__.__name__)

- Mô hình không dự đoán sai mẫu nào

## 📈 (11) So sánh các mô hình

In [15]:
import pandas as pd

# Danh sách kết quả so sánh
comparison = []

# Duyệt qua tất cả các tổ hợp tham số được thử nghiệm trong GridSearchCV
for i in range(len(grid.cv_results_['params'])):
    params = grid.cv_results_['params'][i]
    score = grid.cv_results_['mean_test_score'][i]
    
    # Lấy tên mô hình
    model_name = params['model'].__class__.__name__
    
    # Lấy các tham số của riêng mô hình (bỏ 'model' ra để không hiển thị đối tượng mô hình)
    model_params = {k.replace('model__', ''): v for k, v in params.items() if k != 'model'}
    
    comparison.append({
        'Model': model_name,
        'Parameters': model_params,
        'CV Accuracy': round(score, 4)
    })

# Tạo DataFrame từ danh sách
df_models = pd.DataFrame(comparison)

# Danh sách màu cho từng mô hình
model_colors = {
    'GaussianNB': '#FFF2CC',      # vàng nhạt
    'SVC': '#D9EAD3',             # xanh lục nhạt
    'DecisionTreeClassifier': '#F4CCCC',  # hồng nhạt
    'RandomForestClassifier': '#D0E0E3',  # xanh lam nhạt
    'MLPClassifier': '#EAD1DC'    # tím nhạt
}

def highlight_model(row):
    color = model_colors.get(row['Model'], '#FFFFFF')
    return [f'background-color: {color}'] * len(row)

# Áp dụng tô màu theo từng dòng
styled_df = df_models.style.apply(highlight_model, axis=1)

# Hiển thị bảng với màu
print("📊 So sánh các mô hình và tham số:")
display(styled_df)

📊 So sánh các mô hình và tham số:


,Model,Parameters,CV Accuracy
0,GaussianNB,{},1.000000
1,SVC,"{'C': 0.1, 'kernel': 'linear'}",1.000000
2,SVC,"{'C': 0.1, 'kernel': 'rbf'}",0.993500
3,SVC,"{'C': 1, 'kernel': 'linear'}",1.000000
4,SVC,"{'C': 1, 'kernel': 'rbf'}",0.996800
5,SVC,"{'C': 10, 'kernel': 'linear'}",1.000000
6,SVC,"{'C': 10, 'kernel': 'rbf'}",0.996800
7,DecisionTreeClassifier,"{'criterion': 'gini', 'max_depth': 3}",1.000000
8,DecisionTreeClassifier,"{'criterion': 'gini', 'max_depth': 5}",1.000000
9,DecisionTreeClassifier,"{'criterion': 'gini', 'max_depth': None}",1.000000


- Tuy các mô hình khác cũng có chỉ số accuracy là 100%. Nhưng do GaussNB là mô hình đạt 100% accuracy đầu tiên nên được chọn là best model.
- Mô hình có chỉ số thấp nhất là SVC với tham số {'C': 0.1, 'kernel': 'rbf'} với accuracy là 99.3%